### PR TITLE
Issue 6425: Continuous BookKeeper Log Verification

### LINKED ISSUE #6425: Continuous BookKeeper Log Verification
#### Problem Description
We want to be able to continuously detect anomalies, such as data corruption, from BookKeeper logs.  

#### Problem Location
Segment Store

#### Suggestions for an Improvement
We can create a new admin CLI command to perform a periodic check for `DataCorruptionException` on the available segment containers. The segment container recovery can be implemented based on the existing `container recover` command.

### PR DESCRIPTION
**Pull request by RaulGracia**
**Change log description**  
Adds a new command for performing continuous recovery of all Segment Containers to verify the correctness of logs.

**Purpose of the change**  
Fixes #6425.

**What the code does**  
This PR introduces a new Admin CLI command, namely `ContainerContinuousRecoveryCommand`, that extends from `ContainerRecoverCommand`. The purpose of this new command is to benefit from the existing Segment Container recovery command to do be executed periodically on all Segment Containers, so we can detect failures in the container recovery process continuously until either the number of recovery iterations is met or there is a failure in the recovery process of a Segment Container. One of the most interesting use cases of this new command is to detect such failures when are related to data corruption issues, so we can fix them proactively.

**How to verify it**  
New tests have been added.

Signed-off-by: Raúl Gracia <raul.gracia@emc.com>

### CODE REVIEW

File: cli/admin/src/main/java/io/pravega/cli/admin/bookkeeper/ContainerContinuousRecoveryCommand.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+            // If we reach the desired number of iterations or a failure has happened, release the latch and return.
+            if (counter.get() >= numberOfRuns || failed.get()) {
+                output("Finishing recovery run iterations (failed? " + failed.get() + ")");
+                latch.release();
+                return;
+            }
+            // Recover all the Segment Containers sequentially.
+            for (int i = 0; i < getServiceConfig().getContainerCount(); i++) {
+                try {
+                    super.performRecovery(i);
```
[the remark this answers is not in the document]
Comment (RaulGracia) [reply cli/admin/src/main/java/io/pravega/cli/admin/bookkeeper/ContainerContinuousRecoveryCommand.java:50]: True, this was too verbose. Made the `DebugRecoveryProcessor` callbacks to be overridable, and this command overrides the parent class callbacks with ones that do not print anything (just when something fails).

File: cli/admin/src/main/java/io/pravega/cli/admin/bookkeeper/ContainerContinuousRecoveryCommand.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+                    output("Problem recovering container " + i + ", terminating execution of command.");
+                    failed.set(true);
+                    e.printStackTrace();
+                    break;
+                }
+            }
+            // Increment recovery iterations counter.
+            counter.getAndIncrement();
+        };
+        ScheduledFuture<?> future = getCommandArgs().getState().getExecutor().scheduleAtFixedRate(containerRecoveryIteration,
```
[the remark this answers is not in the document]
Comment (RaulGracia) [reply cli/admin/src/main/java/io/pravega/cli/admin/bookkeeper/ContainerContinuousRecoveryCommand.java]: Ok, used `Futures.loop` instead.

**Comment by RaulGracia** [comment]
@tewit @andreipaduroiu I have tested this command on a real cluster and it works (without being too verbose now):
```
> container continuous-recover 10 10
Recovery complete: 0 DataFrame(s) containing 0 Operation(s).
Completed recovery of container: 0
Recovery complete: 0 DataFrame(s) containing 0 Operation(s).
Completed recovery of container: 1
Recovery complete: 0 DataFrame(s) containing 0 Operation(s).
Completed recovery of container: 2
Recovery complete: 0 DataFrame(s) containing 0 Operation(s).
Completed recovery of container: 3
Recovery complete: 0 DataFrame(s) containing 0 Operation(s).
Completed recovery of container: 4
Recovery complete: 0 DataFrame(s) containing 0 Operation(s).
Completed recovery of container: 5
Recovery complete: 0 DataFrame(s) containing 0 Operation(s).
Completed recovery of container: 6
Recovery complete: 0 DataFrame(s) containing 0 Operation(s).
Completed recovery of container: 7
Recovery complete: 0 DataFrame(s) containing 0 Operation(s).
Completed recovery of container: 8
Recovery complete: 0 DataFrame(s) containing 0 Operation(s).
Completed recovery of container: 9
Recovery complete: 0 DataFrame(s) containing 0 Operation(s).
Completed recovery of container: 10
Recovery complete: 0 DataFrame(s) containing 0 Operation(s).
Completed recovery of container: 11
Recovery complete: 0 DataFrame(s) containing 0 Operation(s).
Completed recovery of container: 12
Recovery complete: 0 DataFrame(s) containing 0 Operation(s).
Completed recovery of container: 13
Recovery complete: 0 DataFrame(s) containing 0 Operation(s).
Completed recovery of container: 14
Recovery complete: 0 DataFrame(s) containing 0 Operation(s).
Completed recovery of container: 15
Recovery complete: 0 DataFrame(s) containing 0 Operation(s).
Completed recovery of container: 16
Recovery complete: 0 DataFrame(s) containing 0 Operation(s).
Completed recovery of container: 17
Recovery complete: 0 DataFrame(s) containing 0 Operation(s).
Completed recovery of container: 18
Recovery complete: 0 DataFrame(s) containing 0 Operation(s).
Completed recovery of container: 19
Recovery complete: 0 DataFrame(s) containing 0 Operation(s).
Completed recovery of container: 20
Recovery complete: 0 DataFrame(s) containing 0 Operation(s).
Completed recovery of container: 21
Recovery complete: 0 DataFrame(s) containing 0 Operation(s).
Completed recovery of container: 22
Recovery complete: 0 DataFrame(s) containing 0 Operation(s).
Completed recovery of container: 23
Recovery complete: 0 DataFrame(s) containing 0 Operation(s).
Completed recovery of container: 0
Recovery complete: 0 DataFrame(s) containing 0 Operation(s).
Completed recovery of container: 1
Recovery complete: 0 DataFrame(s) containing 0 Operation(s).
Completed recovery of container: 2
Recovery complete: 0 DataFrame(s) containing 0 Operation(s).
Completed recovery of container: 3
Recovery complete: 0 DataFrame(s) containing 0 Operation(s).
Completed recovery of container: 4
Recovery complete: 0 DataFrame(s) containing 0 Operation(s).
Completed recovery of container: 5
Recovery complete: 0 DataFrame(s) containing 0 Operation(s).
```

**Comment by RaulGracia** [comment]
@andreipaduroiu did the changes to make the new command less verbose as the regular `container recover` command.

This is an experiment on a real cluster inducing in purpose a `DataCorruptionException` on container 23. As you can see, i) the output is very small on successful recoveries, ii) we see the stacktrace of the exception if a problem happens, iii) the command stops when the first recovery fails.
```
> container continuous-recover 3 3
Recovery complete: 196 DataFrame(s) containing 199 Operation(s).
Completed recovery of container: 0
Recovery complete: 277 DataFrame(s) containing 282 Operation(s).
Completed recovery of container: 1
Recovery complete: 276 DataFrame(s) containing 281 Operation(s).
Completed recovery of container: 2
Recovery complete: 186 DataFrame(s) containing 189 Operation(s).
Completed recovery of container: 3
Recovery complete: 36 DataFrame(s) containing 40 Operation(s).
Completed recovery of container: 4
Recovery complete: 185 DataFrame(s) containing 191 Operation(s).
Completed recovery of container: 5
Recovery complete: 50 DataFrame(s) containing 60 Operation(s).
Completed recovery of container: 6
Recovery complete: 190 DataFrame(s) containing 194 Operation(s).
Completed recovery of container: 7
Recovery complete: 20 DataFrame(s) containing 23 Operation(s).
Completed recovery of container: 8
Recovery complete: 203 DataFrame(s) containing 222 Operation(s).
Completed recovery of container: 9
Recovery complete: 115 DataFrame(s) containing 132 Operation(s).
Completed recovery of container: 10
Recovery complete: 294 DataFrame(s) containing 298 Operation(s).
Completed recovery of container: 11
Recovery complete: 265 DataFrame(s) containing 269 Operation(s).
Completed recovery of container: 12
Recovery complete: 196 DataFrame(s) containing 201 Operation(s).
Completed recovery of container: 13
Recovery complete: 237 DataFrame(s) containing 240 Operation(s).
Completed recovery of container: 14
Recovery complete: 265 DataFrame(s) containing 269 Operation(s).
Completed recovery of container: 15
Recovery complete: 190 DataFrame(s) containing 194 Operation(s).
Completed recovery of container: 16
Recovery complete: 282 DataFrame(s) containing 288 Operation(s).
Completed recovery of container: 17
Recovery complete: 309 DataFrame(s) containing 351 Operation(s).
Completed recovery of container: 18
Recovery complete: 277 DataFrame(s) containing 282 Operation(s).
Completed recovery of container: 19
Recovery complete: 219 DataFrame(s) containing 223 Operation(s).
Completed recovery of container: 20
Recovery complete: 112 DataFrame(s) containing 115 Operation(s).
Completed recovery of container: 21
Recovery complete: 249 DataFrame(s) containing 252 Operation(s).
Completed recovery of container: 22
io.pravega.segmentstore.server.DataCorruptionException: Unable to update metadata for Log Operation 'DeleteSegmentOperation: SequenceNumber = 74, SegmentId = 897, SegmentId = 897'.
        at io.pravega.segmentstore.server.logs.RecoveryProcessor.recoverOperation(RecoveryProcessor.java:194)
        at io.pravega.segmentstore.server.logs.DebugRecoveryProcessor.recoverOperation(DebugRecoveryProcessor.java:119)
        at io.pravega.segmentstore.server.logs.RecoveryProcessor.recoverAllOperations(RecoveryProcessor.java:167)
        at io.pravega.segmentstore.server.logs.RecoveryProcessor.performRecovery(RecoveryProcessor.java:103)
        at io.pravega.segmentstore.server.logs.DebugRecoveryProcessor.performRecovery(DebugRecoveryProcessor.java:51)
        at io.pravega.cli.admin.bookkeeper.ContainerRecoverCommand.performRecovery(ContainerRecoverCommand.java:79)
        at io.pravega.cli.admin.bookkeeper.ContainerContinuousRecoveryCommand.lambda$execute$0(ContainerContinuousRecoveryCommand.java:49)
        at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(Unknown Source)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
        at io.pravega.common.concurrent.ThreadPoolScheduledExecutorService$ScheduledRunnable.run(ThreadPoolScheduledExecutorService.java:189)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.base/java.lang.Thread.run(Unknown Source)
Caused by: io.pravega.segmentstore.server.logs.MetadataUpdateException: [Container 23] No metadata entry exists for Segment Id 897..
        at io.pravega.segmentstore.server.logs.ContainerMetadataUpdateTransaction.getSegmentUpdateTransaction(ContainerMetadataUpdateTransaction.java:557)
        at io.pravega.segmentstore.server.logs.ContainerMetadataUpdateTransaction.preProcessOperation(ContainerMetadataUpdateTransaction.java:337)
        at io.pravega.segmentstore.server.logs.OperationMetadataUpdater.preProcessOperation(OperationMetadataUpdater.java:230)
        at io.pravega.segmentstore.server.logs.RecoveryProcessor.recoverOperation(RecoveryProcessor.java:190)
        ... 12 more
Recovery FAILED: 74 DataFrame(s) containing 74 Operation(s) were able to be recovered.
io.pravega.segmentstore.server.DataCorruptionException: Unable to update metadata for Log Operation 'DeleteSegmentOperation: SequenceNumber = 74, SegmentId = 897, SegmentId = 897'.
        at io.pravega.segmentstore.server.logs.RecoveryProcessor.recoverOperation(RecoveryProcessor.java:194)
        at io.pravega.segmentstore.server.logs.DebugRecoveryProcessor.recoverOperation(DebugRecoveryProcessor.java:119)
        at io.pravega.segmentstore.server.logs.RecoveryProcessor.recoverAllOperations(RecoveryProcessor.java:167)
        at io.pravega.segmentstore.server.logs.RecoveryProcessor.performRecovery(RecoveryProcessor.java:103)
        at io.pravega.segmentstore.server.logs.DebugRecoveryProcessor.performRecovery(DebugRecoveryProcessor.java:51)
        at io.pravega.cli.admin.bookkeeper.ContainerRecoverCommand.performRecovery(ContainerRecoverCommand.java:79)
        at io.pravega.cli.admin.bookkeeper.ContainerContinuousRecoveryCommand.lambda$execute$0(ContainerContinuousRecoveryCommand.java:49)
        at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(Unknown Source)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
        at io.pravega.common.concurrent.ThreadPoolScheduledExecutorService$ScheduledRunnable.run(ThreadPoolScheduledExecutorService.java:189)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.base/java.lang.Thread.run(Unknown Source)
Caused by: io.pravega.segmentstore.server.logs.MetadataUpdateException: [Container 23] No metadata entry exists for Segment Id 897..
        at io.pravega.segmentstore.server.logs.ContainerMetadataUpdateTransaction.getSegmentUpdateTransaction(ContainerMetadataUpdateTransaction.java:557)
        at io.pravega.segmentstore.server.logs.ContainerMetadataUpdateTransaction.preProcessOperation(ContainerMetadataUpdateTransaction.java:337)
        at io.pravega.segmentstore.server.logs.OperationMetadataUpdater.preProcessOperation(OperationMetadataUpdater.java:230)
        at io.pravega.segmentstore.server.logs.RecoveryProcessor.recoverOperation(RecoveryProcessor.java:190)
        ... 12 more
Problem recovering container 23, terminating execution of command.
io.pravega.segmentstore.server.DataCorruptionException: Unable to update metadata for Log Operation 'DeleteSegmentOperation: SequenceNumber = 74, SegmentId = 897, SegmentId = 897'.
        at io.pravega.segmentstore.server.logs.RecoveryProcessor.recoverOperation(RecoveryProcessor.java:194)
        at io.pravega.segmentstore.server.logs.DebugRecoveryProcessor.recoverOperation(DebugRecoveryProcessor.java:119)
        at io.pravega.segmentstore.server.logs.RecoveryProcessor.recoverAllOperations(RecoveryProcessor.java:167)
        at io.pravega.segmentstore.server.logs.RecoveryProcessor.performRecovery(RecoveryProcessor.java:103)
        at io.pravega.segmentstore.server.logs.DebugRecoveryProcessor.performRecovery(DebugRecoveryProcessor.java:51)
        at io.pravega.cli.admin.bookkeeper.ContainerRecoverCommand.performRecovery(ContainerRecoverCommand.java:79)
        at io.pravega.cli.admin.bookkeeper.ContainerContinuousRecoveryCommand.lambda$execute$0(ContainerContinuousRecoveryCommand.java:49)
        at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(Unknown Source)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
        at io.pravega.common.concurrent.ThreadPoolScheduledExecutorService$ScheduledRunnable.run(ThreadPoolScheduledExecutorService.java:189)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.base/java.lang.Thread.run(Unknown Source)
Caused by: io.pravega.segmentstore.server.logs.MetadataUpdateException: [Container 23] No metadata entry exists for Segment Id 897..
        at io.pravega.segmentstore.server.logs.ContainerMetadataUpdateTransaction.getSegmentUpdateTransaction(ContainerMetadataUpdateTransaction.java:557)
        at io.pravega.segmentstore.server.logs.ContainerMetadataUpdateTransaction.preProcessOperation(ContainerMetadataUpdateTransaction.java:337)
        at io.pravega.segmentstore.server.logs.OperationMetadataUpdater.preProcessOperation(OperationMetadataUpdater.java:230)
        at io.pravega.segmentstore.server.logs.RecoveryProcessor.recoverOperation(RecoveryProcessor.java:190)
        ... 12 more
Completed continuous Segment Container recovery command.
```

**Review by andreipaduroiu — APPROVED** [review]

**Pull request merged**